truffle/src/main/java/org/jruby/truffle/parser/BodyTranslator.java

```diff
@@ -2591,11 +2592,33 @@ public RubyNode visitOpAsgnAndNode(OpAsgnAndParseNode node) {
 
     @Override
     public RubyNode visitOpAsgnConstDeclNode(OpAsgnConstDeclParseNode node) {
-        // TODO (eregon, 7 Nov. 2016): Is there any semantic difference?
-        if ("&&".equals(node.getOperator())) {
-            return visitOpAsgnAndNode(new OpAsgnAndParseNode(node.getPosition(), node.getFirstNode(), node.getSecondNode()));
-        } else {
-            return visitOpAsgnOrNode(new OpAsgnOrParseNode(node.getPosition(), node.getFirstNode(), node.getSecondNode()));
+        RubyNode lhs = node.getFirstNode().accept(this);
+        RubyNode rhs = node.getSecondNode().accept(this);
+
+        if (!(rhs instanceof WriteConstantNode)) {
+            rhs = ((ReadConstantNode) lhs).makeWriteNode(rhs);
+        }
+
+        switch (node.getOperator()) {
+            case "&&": {
+                return visitOpAsgnAndNode(node, lhs, rhs);
+            }
+
+            case "||": {
+                final RubyNode defined = new DefinedNode(context, translateSourceSection(source, lhs.getRubySourceSection()), lhs);
+                lhs = new AndNode(defined, lhs);
+
+                return visitOpAsgnOrNode(node, lhs, rhs);
+            }
+
+            default: {
+                final RubySourceSection sourceSection = translate(node.getPosition());
+                final RubyCallNodeParameters callParameters = new RubyCallNodeParameters(context, sourceSection.toSourceSection(source), lhs, node.getOperator(), null, new RubyNode[] { rhs }, false, true, false, false, false);
```

Comment:
There is a slightly shorter version if the last 3 boolean are false for RubyCallNodeParameters.